### DISCUSSION TITLE
MinIO

### DISCUSSION THREAD

**Discussion by ivan-codes-foss** [operatorai/modelstore#183] [Ideas] [answered]
Consider adding support for MinIO

**Comment by nlathia** [comment]
@ivan-codes-foss It looks like the `minio` Python client would fit quite well with `modelstore`. Is this something that you are currently using? (I'm not very familiar with MinIO myself)

https://docs.min.io/docs/python-client-quickstart-guide.html

**Reply by lenoqt** [reply]
As mentioned in #212 I tried using AWSStorage since boto3 can be used with MinIO S3 buckets, the only thing the signature version has to be s3v4.

```python
class MinIOStorage(AWSStorage):
    def validate(self) -> bool:
        logger.debug("Querying for buckets with prefix=%s...", self.bucket_name)
        try:
            resource = boto3.resource('s3',endpoint_url="http://192.168.0.9:32141",
                                      aws_access_key_id=os.environ["MINIO_ACCESS_KEY"],
                                      aws_secret_access_key=os.environ["MINIO_SECRET_ACCESS_KEY"],
                                      config=boto3.session.Config(signature_version="s3v4"), verify=False)
            resource.meta.client.head_bucket(Bucket=self.bucket_name)
            return True
        except ClientError:
            logger.error("Unable to access bucket: %s", self.bucket_name)
            return False
           
class NModelStorage(ModelStore):
    @classmethod
    def from_minio_s3(
        self,
        bucket_name: Optional[str] = None,
        region = None,
        root_prefix: Optional[str] = None,
    ) -> "ModelStore":
        if not BOTO_EXISTS:
            raise ModelNotFoundException("boto 3 is not installed!")
        return ModelStore(storage=MinIOStorage(bucket_name=bucket_name))

m = NModelStorage.from_minio_s3(bucket_name='ml-metadata') 
m.upload("testdomain", model=model) 
```

But this didn't work, I honestly don't know if I am doing something wrong here when trying to upload a Transformer model.

**Reply by nlathia** [reply]
What error did you get when uploading the transformer model?

**Reply by lenoqt** [reply]
On my app, it goes on 5 retries then fails to reach host but running this same piece of code on a jupyter notebook worked like a charm, so my guess is that the most straight forward to make it work it is to modify the ```AWSStorage``` implementation to allow changes in the ```boto3.resource``` signature to s3v4.

**Comment by nlathia** [comment]
@lenoqt @ivan-codes-foss I've started adding this into `modelstore` - the PR is not finished, but [the draft is here](https://github.com/operatorai/modelstore/pull/214). Will wrap it up soon, and then this can be added into the next release of `modelstore`.

**Comment by nlathia** [comment, marked as answer]
Update - this is now merged, and modelstore + MinIO (with an s3 bucket) runs as part of this repo's PR integration tests! This will go out in `modelstore==0.0.77`

- https://github.com/operatorai/modelstore/pull/214

**Reply by lenoqt** [reply]
Thank you for this! I will test it whenever I get a chance.
